src/Lean/Compiler/IR/LLVMBindings.lean

```diff
@@ -0,0 +1,301 @@
+/-
+Copyright (c) 2022 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Siddharth Bhat
+
+This file defines bindings for LLVM. The main mechanisms
+are to:
+- Mirror the values of C enumerations.
+- Create opaque values for LLVM's pointer based API
+- Write bindings that are `IO` based, which mutate the underlying in-memory
+  data structures to build IR.
+-/
+
+namespace LLVM
+
+namespace CodeGenFileType
+-- https://github.com/llvm/llvm-project/blob/c3e073bcbdc523b0f758d44a89a6333e38bff863/llvm/include/llvm-c/TargetMachine.h#L64
+def AssemblyFile : UInt64 := 0
+def ObjectFile : UInt64 := 1
+end CodeGenFileType
+
+namespace IntPredicate
+-- https://github.com/llvm/llvm-project/blob/c3e073bcbdc523b0f758d44a89a6333e38bff863/llvm/include/llvm-c/Core.h#L290
+def EQ : UInt64 := 32
+def NE : UInt64 := EQ + 1
+def UGT : UInt64 := NE + 1
+end IntPredicate
+
+inductive Ty where
+| int: Nat → Ty
+| float: Nat → Ty
+| ptr: Ty → Ty
+
+def Ty.i32: Ty := Ty.int 32
+def Ty.i1: Ty := Ty.int 1
+def Ty.i8: Ty := Ty.int 8
+def Ty.i8ptr: Ty := .ptr <| Ty.i8
+
+structure BasicBlock where
+structure Context where
+structure Module where
+structure Builder where
+structure LLVMType where
+structure Value where
+structure MemoryBuffer where
+structure Target where
+structure TargetMachine where
```

Comment:
If these types exist merely for building type-safe interfaces, they should probably be opaque as well